File: l33tcode/unique-binary-search-trees.py
from functools import cache


class Solution:
    def numTrees(self, n: int) -> int:
        @cache
        def dfs(nodes: int) -> int:
            if nodes == 0:
                return 1

            count = 0
            for split in range(nodes):
                left_subtrees = dfs(split)
                right_subtrees = dfs(nodes - split - 1)
                count += left_subtrees * right_subtrees

            return count

        return dfs(n)

File: l33tcode/unique-binary-search-trees.go
package main

func numTrees(n int) int {
	dp := make([]int, n+1)
	dp[0] = 1

	for num := range make([]int, n) {
		elements := num + 1
		for split := range make([]int, elements) {
			dp[elements] += dp[split] * dp[elements-split-1]
		}
	}

	return dp[n]
}
